### ISSUE TITLE
Hot-Restarts are stuck in flutter web after updating flutter to v3.7.1 and 3.7.3

### ISSUE THREAD

**Issue by mhbdev** [flutter/flutter#120310]
## Steps to Reproduce

I upgraded flutter to the latest version 3.7 on stable channel and this issue appeared. You can see `flutter doctor -v` upfront.

[√] Flutter (Channel stable, 3.7.1, on Microsoft Windows [Version 10.0.19045.2486], locale en-US)
    • Flutter version 3.7.1 on channel stable at C:\FlutterSdk\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7048ed95a5 (7 days ago), 2023-02-01 09:07:31 -0800
    • Engine revision 800594f1f4
    • Dart version 2.19.1
    • DevTools version 2.20.1
    • Pub download mirror https://pub.flutter-io.cn

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
    • Android SDK at C:\AndroidSdk
    • Platform android-33, build-tools 33.0.1
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.3.5)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.3.32922.545
    • Windows 10 SDK version 10.0.20348.0

[√] Android Studio (version 2021.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[√] VS Code (version 1.74.3)
    • VS Code at C:\Users\MonoBit\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.58.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19045.2486]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 109.0.5414.120
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 108.0.1462.54

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

Here is `flutter run -d chrome --verbose` after entering 'R' to hotRestart I get the log below

```
[        ] DwdsVmClient: Attempting a hot restart
[        ] DwdsVmClient: Attempting to disable breakpoints and resume the isolate
[        ] DwdsVmClient: Successfully disabled breakpoints and resumed the isolate
[        ] DwdsVmClient: Attempting to get execution context ID.
[        ] DwdsVmClient: Got execution context ID.
[        ] DwdsVmClient: Issuing $dartHotRestartDwds request
[   +1 ms] ChromeProxyService: Destroying isolate
[        ] BatchedExpressionEvaluator: Closed
[  +31 ms] DwdsVmClient: $dartHotRestartDwds request complete.
[        ] DwdsVmClient: Waiting for Isolate Start event.
[   +5 ms] ChromeProxyService: Destroying isolate
[        ] ChromeProxyService: Destroying isolate
```

And every time I am seeing this error in my chrome console log
* Refused to execute script from 'http://localhost:51565/user/main.dart.js' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.

**Comment by mhbdev** [comment]
I think something is wrong here in the last `then` but even if so, That is so weird because it was working perfectly fine before.

```javascript
 window.addEventListener('load', function() {
      var loading = document.querySelector('#loading');
      _flutter.loader.loadEntrypoint({
        serviceWorker: {
          serviceWorkerVersion: serviceWorkerVersion,
        }
      }).then(function(engineInitializer) {
        loading.classList.add('main_done');
        return engineInitializer.initializeEngine();
      }).then(function(appRunner) {
        loading.classList.add('init_done');
        return appRunner.runApp();
      }).then(function(app) {
        // Wait a few milliseconds so users can see the "zoooom" animation
        // before getting rid of the "loading" div.
        window.setTimeout(function() {
          loading.remove();
        }, 200);
      });
    });
```

**Comment by darshankawar** [comment]
@mhbdev 
This probably should be fixed by https://github.com/flutter/flutter/pull/118684, so I suggest you to switch to latest master and try the scenario again to see what you get.

**Comment by taufiq-husain** [comment]
Hey, I have had the exact same issue for 2+ weeks. I constantly upgrade flutter to the newest versions. @darshankawar my problem is this:

I open a project in VSCode. I run it using the terminal (for flutter web i.e. `flutter run -d chrome`) or I run it by pressing ^F5. I see my app on the web, it's working fine. Then when I change the code a little (even adding a space somewhere) and press save, VS code says "Performing hot restart..." and then the web app shows the loading icon and never finishes loading. So I try to press reload/restart again and I get this error: `Reload already in progress, ignoring request`

Here's my `flutter doctor -v` results:

`
taufiqhusain@Taufiqs-Airbook kokkatoo % flutter doctor -v                                             
[✓] Flutter (Channel stable, 3.7.3, on macOS 13.1 22C65 darwin-arm64, locale en-US)
    • Flutter version 3.7.3 on channel stable at /Users/taufiqhusain/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9944297138 (31 hours ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/taufiqhusain/Library/Android/sdk
    • Platform android-33, build-tools 32.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.75.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 109.0.5414.119

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
`

Any idea how to solve this?

**Comment by mhbdev** [comment]
> @mhbdev This probably should be fixed by #118684, so I suggest you to switch to latest master and try the scenario again to see what you get.

I upgraded flutter to version 3.7.3 and it seems the problem is solved with nested routes

**Comment by mhbdev** [comment]
> @mhbdev This probably should be fixed by #118684, so I suggest you to switch to latest master and try the scenario again to see what you get.

It is not fixed yet! I am on the latest release of flutter version 3.7.3 and the issue still exists

**Comment by mhbdev** [comment]
I found out that this is happening in debug Mode when I am using nested routing and there is a usage of usePathUrlStrategy therefore I changed this and rewrote that, in a new way :

```dart
if(kReleaseMode) {
    usePathUrlStrategy();
}
```

**Comment by taufiq-husain** [comment]
@mhbdev does it still work on release mode if u refresh the page? Like if you refresh the browser a few times, it doesn't get stuck like that anymore?

**Comment by Quantentunnel** [comment]
> > @mhbdev This probably should be fixed by #118684, so I suggest you to switch to latest master and try the scenario again to see what you get.
> 
> It is not fixed yet! I am on the latest release of flutter version 3.7.3 and the issue still exists

I can confirm that I still have the error with this version as well.

**Comment by Quantentunnel** [comment]
> I found out that this is happening in debug Mode when I am using nested routing and there is a usage of usePathUrlStrategy therefore I changed this and rewrote that, in a new way :
> 
> ```dart
> if(kReleaseMode) {
>     usePathUrlStrategy();
> }
> ```

I did not know what the variable "kReleaseMode" contains. So I commented out setPathUrlStrategy(); and it seems to have a positive effect.
The question for me is whether the error is in Flutter itself or the extension url_strategy.

**Comment by darshankawar** [comment]
@mhbdev 
I am unable to replicate this using latest stable 3.7.3 using a demo app. Although I am not using VSCode, I tried the same using AS and running the web app on mac + chrome.

Can you try the same using terminal / command line directly instead of through VSCode and see if you still get same behavior or not ?

**Comment by mhbdev** [comment]
> > I found out that this is happening in debug Mode when I am using nested routing and there is a usage of usePathUrlStrategy therefore I changed this and rewrote that, in a new way :
> > ```dart
> > if(kReleaseMode) {
> >     usePathUrlStrategy();
> > }
> > ```
> 
> I did not know what the variable "kReleaseMode" contains. So I commented out setPathUrlStrategy(); and it seems to have a positive effect. The question for me is whether the error is in Flutter itself or the extension url_strategy.

I used kReleaseMode check to disable `setPathUrlStrategy();` on debug mode therefore the error disappears when there is no urlStrategy. commenting that out will do the same.

**Comment by mhbdev** [comment]
> @mhbdev I am unable to replicate this using latest stable 3.7.3 using a demo app. Although I am not using VSCode, I tried the same using AS and running the web app on mac + chrome.
> 
> Can you try the same using terminal / command line directly instead of through VSCode and see if you still get same behavior or not ?

I am also using Android Studio and I am on Windows. I tried the terminal and still the same event happening. The only way to bypass that hot-restart hanging problem is to refresh the browser manually.

**Comment by mhbdev** [comment]
> @mhbdev does it still work on release mode if u refresh the page? Like if you refresh the browser a few times, it doesn't get stuck like that anymore?

Yes I tried that out and this kReleaseMode thing is only used to prevent urlPathStrategy in debugMode you can also use kDebugMode to neutralize its functionality. I am using this .htaccess file to handle endpoints on my server. This might be helpful for release mode.

This snippet will redirect all incoming endpoints to `/#/{requestedEndPoint}` and it is working perfectly fine in release mode.

```
<IfModule mod_rewrite.c>
    <IfModule mod_negotiation.c>
        Options -MultiViews -Indexes
    </IfModule>

    RewriteEngine On

    # Handle Authorization Header
    RewriteCond %{HTTP:Authorization} .
    RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]

    # Redirect Trailing Slashes If Not A Folder...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_URI} (.+)/$
    RewriteRule ^ %1 [L,R=301]

    # Handle Front Controller...
    RewriteCond %{REQUEST_FILENAME} !-d
    RewriteCond %{REQUEST_FILENAME} !-f
    RewriteRule ^ index.html [L]
</IfModule>
```

**Comment by darshankawar** [comment]
> > It is not fixed yet! I am on the latest release of flutter version 3.7.3 and the issue still exists
> 
> I can confirm that I still have the error with this version as well.

@mhbdev @Quantentunnel I meant to verify on latest master version to see if [this](https://github.com/flutter/flutter/pull/118684) fix has any effect on the behavior you are seeing. 

`flutter channel master`
`flutter upgrade`

**Comment by taufiq-husain** [comment]
> > > It is not fixed yet! I am on the latest release of flutter version 3.7.3 and the issue still exists
> > 
> > 
> > I can confirm that I still have the error with this version as well.
> 
> @mhbdev @Quantentunnel I meant to verify on latest master version to see if [this](https://github.com/flutter/flutter/pull/118684) fix has any effect on the behavior you are seeing.
> 
> `flutter channel master` `flutter upgrade`

Seems to be working for me now. It's not in stable channel yet?

**Comment by mhbdev** [comment]
> > > It is not fixed yet! I am on the latest release of flutter version 3.7.3 and the issue still exists
> > 
> > 
> > I can confirm that I still have the error with this version as well.
> 
> @mhbdev @Quantentunnel I meant to verify on latest master version to see if [this](https://github.com/flutter/flutter/pull/118684) fix has any effect on the behavior you are seeing.
> 
> `flutter channel master` `flutter upgrade`

That is solving the problem. This is really weird that this kind of issues are happening in stable channel latest releases. Thanks for your help.

**Comment by darshankawar** [comment]
The fix is being cherry picked to stable: https://github.com/flutter/flutter/issues/120197 so please subscribe to this issue. Once this is merged, a new hotfix stable version will be available with the fix.

Closing this issue since this is fixed in master.

**Comment by tkreuder** [comment]
![Screenshot 2023-02-15 at 08 06 35](https://user-images.githubusercontent.com/385110/218956739-3d4f507e-eb79-4191-a03c-97e041a70c81.png)
I am still stuck in this state on hot-restart.

![image](https://user-images.githubusercontent.com/385110/218956888-b02cb712-ad4b-418c-9903-aeee7c2a361e.png)


> I think something is wrong here in the last `then` but even if so, That is so weird because it was working perfectly fine before.
> 
> ```js
>  window.addEventListener('load', function() {
>       var loading = document.querySelector('#loading');
>       _flutter.loader.loadEntrypoint({
>         serviceWorker: {
>           serviceWorkerVersion: serviceWorkerVersion,
>         }
>       }).then(function(engineInitializer) {
>         loading.classList.add('main_done');
>         return engineInitializer.initializeEngine();
>       }).then(function(appRunner) {
>         loading.classList.add('init_done');
>         return appRunner.runApp();
>       }).then(function(app) {
>         // Wait a few milliseconds so users can see the "zoooom" animation
>         // before getting rid of the "loading" div.
>         window.setTimeout(function() {
>           loading.remove();
>         }, 200);
>       });
>     });
> ```
> 
> I think I need to add more details that are noticeable
> 
> 1. I am using auto_route and this case only happens when I am hot-restarting from an inner route like `dashboard/first` or `dashboard/second` which is implemented via AutoTabsRouter.
> 2. By Deleting the endpoints from `localhost:xxxx/dashboard/first` and reloading `localhost:xxxx` the lock of hot-restart will be broken but I am still having the same issue for next hot restarts.

This does not help for me.  It shows the following in the logs and in the browser the loading indicator at the top re-appears, but it is still stuck in that state. I am using  go_router: 3.1.0.
![Screenshot 2023-02-15 at 08 07 03](https://user-images.githubusercontent.com/385110/218958474-7b9ef880-ef24-4e1c-bf3f-d23da111d778.png)

![image](https://user-images.githubusercontent.com/385110/218958448-c9586067-b372-42ce-ad16-333c5e9795fe.png)

**Comment by tkreuder** [comment]
I tried with Visual Studio code and hot-restart works. It does not work with IntelliJ based IDEs. Android Studio 2022.1.1 and IntelliJ 2022.2.3 both have this issue.
![image](https://user-images.githubusercontent.com/385110/218974812-ad002a16-be71-48e7-9e0e-79c08c3a8e91.png)

**Comment by mhbdev** [comment]
I switched to master channel latest release with that fix you mentioned but something odd is happening in my release version. I am getting these errors in chrome console
![image](https://user-images.githubusercontent.com/53684884/219019993-7ae6c7ee-b786-4ca0-91e0-5edd1f57ca90.png)
![image](https://user-images.githubusercontent.com/53684884/219020052-187b5fc0-6afa-4626-a1e9-036f7b9401e0.png)

Should open another issue for this?

**Comment by mhbdev** [comment]
> I switched to master channel latest release with that fix you mentioned but something odd is happening in my release version. I am getting these errors in chrome console ![image](https://user-images.githubusercontent.com/53684884/219019993-7ae6c7ee-b786-4ca0-91e0-5edd1f57ca90.png) ![image](https://user-images.githubusercontent.com/53684884/219020052-187b5fc0-6afa-4626-a1e9-036f7b9401e0.png)
> 
> Should open another issue for this?

I found out that the errors are occurring because of `--dart-define=FLUTTER_WEB_CANVASKIT_URL=/dashboard/canvaskit/` but I have to mention that this parameter was already working fine on the stable channel version 3.7.3. Any idea why is this happening?

**Comment by tkreuder** [comment]
> > I switched to master channel latest release with that fix you mentioned but something odd is happening in my release version. I am getting these errors in chrome console ![image](https://user-images.githubusercontent.com/53684884/219019993-7ae6c7ee-b786-4ca0-91e0-5edd1f57ca90.png) ![image](https://user-images.githubusercontent.com/53684884/219020052-187b5fc0-6afa-4626-a1e9-036f7b9401e0.png)
> > Should open another issue for this?
> 
> I found out that the errors are occurring because of `--dart-define=FLUTTER_WEB_CANVASKIT_URL=/dashboard/canvaskit/` but I have to mention that this parameter was already working fine on the stable channel version 3.7.3. Any idea why is this happening?
> 
> > Note: dashboard is a subfolder in my host so the base-href is `/dashboard/`

Is it not an asset of the flutter toast library? 
[fluttertoast](https://pub.dev/packages/fluttertoast )
[toastify.js](https://github.com/ponnamkarthik/FlutterToast/blob/master/assets/toastify.js)

**Comment by mhbdev** [comment]
> > > I switched to master channel latest release with that fix you mentioned but something odd is happening in my release version. I am getting these errors in chrome console ![image](https://user-images.githubusercontent.com/53684884/219019993-7ae6c7ee-b786-4ca0-91e0-5edd1f57ca90.png) ![image](https://user-images.githubusercontent.com/53684884/219020052-187b5fc0-6afa-4626-a1e9-036f7b9401e0.png)
> > > Should open another issue for this?
> > 
> > 
> > I found out that the errors are occurring because of `--dart-define=FLUTTER_WEB_CANVASKIT_URL=/dashboard/canvaskit/` but I have to mention that this parameter was already working fine on the stable channel version 3.7.3. Any idea why is this happening?
> > > Note: dashboard is a subfolder in my host so the base-href is `/dashboard/`
> 
> Is it not an asset of the flutter toast library? [fluttertoast](https://pub.dev/packages/fluttertoast) [toastify.js](https://github.com/ponnamkarthik/FlutterToast/blob/master/assets/toastify.js)

I think I have to go back to stable channel because there are several issues here in master channel.
by removing `--dart-define` argument the error has been disappeared. I checked my assets folder in release built files and folders and I am sure that my font is in the right place.

![image](https://user-images.githubusercontent.com/53684884/219055449-c8629ce5-3e93-4fcd-b847-e3ba0cd197f0.png)
